### PR TITLE
add support for multi-query planning

### PR DESCRIPTION
**Pull request by mamoll**
### Description

If enabled, the internal state of of planners does not get cleared.
This can be enabled per planner. It only makes sense for roadmap-based
planners (or if the start state does not change). Additionally, for
PRM, PRM*, LazyPRM, and LazyPRM* it is possible (when using OMPL 1.5.0)
to load/save roadmaps from disk upon creation/destruction. This is
set in ompl_planning.yaml like so:

```
planner_configs:
  PersistentPRM:
    type: geometric::PRM
    multi_query_planning_enabled: true
    store_planner_data: true
    load_planner_data: true
    planner_data_path: /tmp/roadmap.graph
```

This commit also fixes a bug for PRM-family planners where they would
always run as optimizing planners and not terminate when the first
solution was found. This occurs when the default optimization objective
is set when none are specified. The fix is to not set an objective
when none are specified.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the MIGRATION.md notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by mamoll** [comment]
@henningkayser, I refactored your code a bit. Let me know if any bugs have slipped in.

File: moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp [hunk 1]
```diff
@@ -85,6 +85,136 @@ struct PlanningContextManager::CachedContexts
...
+  //  Remember which planner instances to store when the destructor is called
+  if (store_planner_data)
+    planner_data_storage_paths_[new_name] = file_path;
+  return planner;
+}
+
+template <typename T>
+inline ompl::base::Planner* ompl_interface::MultiQueryPlannerAllocator::allocatePersistentPlanner(const ob::PlannerData& data)
+{
+  ROS_ERROR_NAMED("planning_context_manager", "Creating a planner instance from PlannerData failed");
```
Comment (rhaschke) [inline moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp]: Error message moved `allocatePlannerImpl`
```suggestion
```

File: moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp [hunk 2]
```diff
@@ -349,8 +349,9 @@ void ompl_interface::ModelBasedPlanningContext::useConfig()
   {
     std::string type = it->second;
     cfg.erase(it);
-    ompl_simple_setup_->setPlannerAllocator(std::bind(spec_.planner_selector_(type), std::placeholders::_1,
-                                                      name_ != getGroupName() ? name_ : "", std::cref(spec_)));
+    const std::string planner_name = getGroupName() + "/" + name_;
```
[the remark this answers is not in the document]
Comment (henningkayser) [reply moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp]: This is required in order to support multiple instances of the same planner for multiple planning groups

File: moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/planning_context_manager.h [hunk 3]
```diff
@@ -41,12 +41,47 @@
...
+  template <typename T>
+  inline ob::Planner* allocatePersistentPlanner(const ob::PlannerData& data);
+
+  // Storing multi-query planners
+  std::map<std::string, ob::PlannerPtr> planners_;
+
+  std::map<std::string, std::string> planner_data_storage_paths_;
+
+  // Store and load planner data
+  ob::PlannerDataStorage storage_;
```
[the remark this answers is not in the document]
Comment (mamoll) [reply moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/planning_context_manager.h]: It's simply a data writer. The `storage_` member is not planner specific. (It is move group specific, though.)

File: moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp [hunk 4]
```diff
@@ -298,44 +299,43 @@ void ompl_interface::ModelBasedPlanningContext::useConfig()
   std::string optimizer;
   ompl::base::OptimizationObjectivePtr objective;
   it = cfg.find("optimization_objective");
-  if (it == cfg.end())
```
[the remark this answers is not in the document]
Comment (mamoll) [reply moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp]: If no optimization objective is set, the optimizing planners will internally set the objective to be path length. The external behavior is not changed by this change. It only changes the buggy behavior of the non-optimizing PRM and LazyPRM planners.

File: moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp [hunk 5]
```diff
@@ -85,6 +85,136 @@ struct PlanningContextManager::CachedContexts
...
+  bool multi_query_planning_enabled = false;
+  if (it != cfg.end())
+  {
+    multi_query_planning_enabled = boost::lexical_cast<bool>(it->second);
+    cfg.erase(it);
+  }
+  if (multi_query_planning_enabled)
+  {
+    // If we already have an instance, use that one
+    if (planners_.count(new_name))
```
[the remark this answers is not in the document]
Comment (mamoll) [reply moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp]: Good catch. Fixed in latest commit.

**Comment by simonschmeisser** [comment]
Just curious and a bit offtopic, can this load and store capability cope with slightly changing environments or especially in our case environments where only a very defined region of the workspace changes but the rest is static? (Eg we do bin-picking from a random box but then usually deposit at a fixed place)
How are attached objects handled, are they checked after the main optimization in a validation step?

**Comment by mamoll** [comment]
> Just curious and a bit offtopic, can this load and store capability cope with slightly changing environments or especially in our case environments where only a very defined region of the workspace changes but the rest is static? (Eg we do bin-picking from a random box but then usually deposit at a fixed place)
How are attached objects handled, are they checked after the main optimization in a validation step?

@simonschmeisser if the environment changes slightly or if the attached objects change, then you should use LazyPRM or LazyPRM*. These two planners will re-validate nodes and edges as needed when trying to find a path. (The initial validity of each node and edge is marked as `VALIDITY_UNKNOWN`.) This should still be better than recreating a roadmap from scratch if the environment only changes slightly.

**Pull request merged**